### PR TITLE
Add st7920 display,

### PR DESCRIPTION
**Pull request by marsjan155**
Added ST7920 LCD graphic display, tested on 128x64 model

## Description:

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#907

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Misiu** [comment]
@marsjan155 I have this display laying around so I can test this, but I need to know how to do that. I have Home Assistant running on Pi4 (standard setup).
@jesserockz can this be cherry-picked to 1.6 beta3? This way we'll be able to use it and test from Home Assistant.

**Comment by jesserockz** [comment]
It cannot be cherry picked until merged into Dev.
You can test a branch on the ha add-on with this add-on config:
```
esphome_version: "marsjan155:st7920"
```

**Comment by Misiu** [comment]
I'll set up dev env and will try to test this. Hopefully, this can be merged into dev and cherry-picked before 1.6 will be released.
I didn't know we can select a specific version of ESPHome like this. Thanks for the tip 🙂

**Comment by mdepa-cl** [comment]
@Misiu It'll be great if someone else confirms it's working

**Comment by marsjan155** [comment]
@Misiu It'll be great if someone else confirms it's working

**Review by jesserockz — CHANGES_REQUESTED** [review]
Since this is an SPIDevice, why does it need a separate `RS` pin? The built in `cs_pin` does this job with `this->enable()` / `this->disable()` and if you need it to be high instead of low, you can invert the pin in yaml.

If this is the case, please remove the rs_pin from the code and the docs.

**Comment by marsjan155** [comment]
@jesserockz right but for st7920 will be always inverted, will not conflict with other SPI devices? Maybe can be inverted programmatically before each transmission?

**Comment by jesserockz** [comment]
So the best place to invert is in the yaml, otherwise a user might invert it, then the code will invert it too.
Which means you just need to update the docs to inform users that the pin should be inverted in their config

**Comment by marsjan155** [comment]
@jesserockz
I found some free time and pushed changes as suggested. It works and please check it

File: esphome/components/st7920/st7920.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  } else if (y >= 64 + 32 && y < 64 + 64) {
+    y -= 64;
+    x += 8;
+  }
+  this->command_(LCD_ADDR | y);  // 6-bit (0..63)
+  this->command_(LCD_ADDR | x);  // 4-bit (0..15)
+}
+
+void HOT ST7920::write_display_data() {
+  byte i, j, b;
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]: `byte` is an Arduino-ism, we prefer standard C++.
```suggestion
  uint8_t i, j, b;
```

File: esphome/components/st7920/st7920.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    for (i = 0; i < 16; i++) {  // 16 bytes from line #32+
+      b = this->buffer_[i + (j + 32) * 16];
+      this->send_(LCD_DATA, b);
+    }
+    this->disable();
+    App.feed_wdt();
+  }
+}
+
+void ST7920::fill(Color color) { memset(this->buffer_, 0, this->get_buffer_length_()); }
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]: This ignores the `color` argument. I presume it should be something like this:
```suggestion
void ST7920::fill(Color color) { 
  memset(this->buffer_, color.is_on() ? 0xFF : 0x00, this->get_buffer_length_()); 
}
```

File: esphome/components/st7920/st7920.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+void ST7920::dump_config() {
+  LOG_DISPLAY("", "ST7920", this);
+  LOG_PIN("  CS Pin: ", this->cs_);
+  ESP_LOGD(TAG, "  Height: %d", this->height_);
+  ESP_LOGD(TAG, "  Width: %d", this->width_);
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]:
```suggestion
  ESP_LOGCONFIG(TAG, "  Height: %d", this->height_);
  ESP_LOGCONFIG(TAG, "  Width: %d", this->width_);
```

File: esphome/components/st7920/st7920.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+float ST7920::get_setup_priority() const { return setup_priority::PROCESSOR; }
+
+void ST7920::update() {
+  this->fill(COLOR_OFF);
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]:
```suggestion
  this->clear();
```

File: esphome/components/st7920/st7920.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  if (this->writer_local_.has_value())  // insert Labda function if available
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]:
```suggestion
  if (this->writer_local_.has_value())  // call lambda function if available
```

File: esphome/components/st7920/st7920.cpp [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    (*this->writer_local_)(*this);
+  this->write_display_data();
+}
+
+void ST7920::loop() {}
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]: This is unnecessary, you can just leave the function out if you don't need it.
```suggestion
```

File: esphome/components/st7920/st7920.cpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+int ST7920::get_height_internal() { return this->height_; }
+
+size_t ST7920::get_buffer_length_() {
+  return size_t(this->get_width_internal()) * size_t(this->get_height_internal()) / 8u;
+}
+
+void HOT ST7920::draw_absolute_pixel_internal(int x, int y, Color color) {
+  if (x >= this->get_width_internal() || x < 0 || y >= this->get_height_internal() || y < 0) {
+    ESP_LOGW(TAG, "Posiotion out of area: %dx%d", x, y);
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]:
```suggestion
    ESP_LOGW(TAG, "Position out of area: %dx%d", x, y);
```

File: esphome/components/st7920/st7920.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  int width = this->get_width_internal() / 8u;
+  if (color.is_on()) {
+    this->buffer_[y * width + x / 8] |= (0x80 >> (x & 7));
+  } else {
+    this->buffer_[y * width + x / 8] &= ~(0x80 >> (x & 7));
+  }
+}
+
+void ST7920::display_init_() {
+  ESP_LOGCONFIG(TAG, "displayInit...");
```
Comment (oxan) [inline esphome/components/st7920/st7920.cpp]:
```suggestion
  ESP_LOGD(TAG, "Initializing display...");
```

File: esphome/components/st7920/st7920.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+#include "esphome/core/component.h"
+#include "esphome/components/display/display_buffer.h"
+#include "esphome/components/spi/spi.h"
+
+namespace esphome {
+namespace st7920 {
+
+/// Turn the pixel ON.
+const Color COLOR_OFF(0, 0, 0, 0);
+const Color COLOR_ON(1, 1, 1, 1);
```
Comment (oxan) [inline esphome/components/st7920/st7920.h]: With my previous suggestion, these aren't used anymore:
```suggestion
```

File: esphome/components/st7920/st7920.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  void set_height(uint16_t height) { this->height_ = height; }
+  void set_width(uint16_t width) { this->width_ = width; }
+
+  // ========== INTERNAL METHODS ==========
+  // (In most use cases you won't need these)
+  void setup() override;
+  void dump_config() override;
+  float get_setup_priority() const override;
+  void update() override;
+  void loop() override;
```
Comment (oxan) [inline esphome/components/st7920/st7920.h]:
```suggestion
```

**Review by oxan — COMMENTED** [review]
I've some suggestions for minor issues (you can commit them all at once from the "Files changed" tab), otherwise this looks good!

**Comment by marsjan155** [comment]
@oxan I'm learning cpp in my free time and learned something new from your review, thanks a lot
only need to pass checks :)

**Comment by oxan** [comment]
> only need to pass checks :)

And check whether it works ;) (Though I'd be surprised if any of my suggestions broke it, but you never know).

**Comment by marsjan155** [comment]
yeah, small changes can be dangerous :) checked also on device - still works

**Review by oxan — APPROVED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**